zookeeper/source/zk_nn_client.cc

```diff
@@ -0,0 +1,952 @@
+#ifndef RDFS_ZKNNCLIENT_CC
+#define RDFS_ZKNNCLIENT_CC
+
+#include "../include/zk_nn_client.h"
+#include "zkwrapper.h"
+#include <iostream>
+#include <sstream>
+#include <ctime>
+#include <chrono>
+#include <sys/time.h>
+#include "zk_lock.h"
+
+#include "hdfs.pb.h"
+#include "ClientNamenodeProtocol.pb.h"
+#include <google/protobuf/message.h>
+#include <ConfigReader.h>
+#include <easylogging++.h>
+
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/algorithm/string.hpp>
+#include "zk_dn_client.h"
+
+namespace zkclient{
+
+	using namespace hadoop::hdfs;
+
+	const std::string ZkNnClient::CLASS_NAME = ": **ZkNnClient** : ";
+
+	ZkNnClient::ZkNnClient(std::string zkIpAndAddress) : ZkClientCommon(zkIpAndAddress) {
+		mkdir_helper( "/", false);
+	}
+
+	ZkNnClient::ZkNnClient(std::shared_ptr <ZKWrapper> zk_in) : ZkClientCommon(zk_in) {
+		mkdir_helper( "/", false);
+	}
+
+	/*
+	 * A simple print function that will be triggered when
+	 * namenode loses a heartbeat
+	 */
+	void notify_delete() {
+		printf("No heartbeat, no childs to retrieve\n");
+	}
+
+	void ZkNnClient::register_watches() {
+
+		int error_code;
+		// TODO: Do we have to free the returned children?
+		std::vector <std::string> children = std::vector <std::string>();
+
+		/* Place a watch on the health subtree */
+
+
+		if (!(zk->wget_children(HEALTH, children, zk->watcher_health_factory(ZkClientCommon::HEALTH_BACKSLASH), nullptr, error_code))) {
+				// TODO: Handle error
+				LOG(ERROR) << CLASS_NAME << "[In register_watchers], wget failed " << error_code;
+			}
+
+		for (int i = 0; i < children.size(); i++) {
+			LOG(INFO) << CLASS_NAME << "[In register_watches] Attaching child to " << children[i] << ", ";
+			std::vector <std::string> ephem = std::vector <std::string>();
+			if(!(zk->wget_children(HEALTH_BACKSLASH + children[i], ephem, zk->watcher_health_child, nullptr, error_code))) {
+				// TODO: Handle error
+				LOG(ERROR) << CLASS_NAME << "[In register_watchers], wget failed " << error_code;
+			}
+			/*
+			   if (ephem.size() > 0) {
+			   LOG(INFO) << CLASS_NAME << "Found ephem " << ephem[0];
+			   } else {
+			   LOG(INFO) << CLASS_NAME << "No ephem found for " << children[i];
+			   }
+			 */
+		}
+	}
+
+	bool ZkNnClient::file_exists(const std::string& path) {
```

Comment:
All internal (probably should be private) ZkNnClient methods should only accept ZKFileSystem prepended paths. They should check for this with asserts or something.